### PR TITLE
Build: Upgrade Jest to 29

### PR DESCRIPTION
**Pull request by IanVS**
Issue:

## What I did

This bumps jest to the latest version, 29.3.1, along with a few other related jest packages.

I needed to break up the config a little, because I started getting validation messages about the "global" config options that I needed to move to the top level.

## How to test

- CI

### CODE REVIEW

**Comment by kasperpeulen** [comment]
@IanVS Just did a quick commit, to see if we can remove run in band with jest 29, to see if we can speed things up again in CI.

File: .circleci/config.yml
Comment (IanVS) [inline .circleci/config.yml:380]: @yannbf and I found that this flag is required in order for the tests to pass in circleCI.  Otherwise, it stalls out after 10 minutes.  With the flag, it completes very quickly.

**Comment by kasperpeulen** [comment]
@IanVS Removing runInBand made it crash indeed, but setting maxWorkers=4 seems to be green:

image

When I initially removed runInBand a couple of weeks ago, I saw a consistent 2.6 improvement. Just comparing those 2 gives 2.4 time improvement.

**Review by kasperpeulen — CHANGES_REQUESTED** [review]
LGTM!

**Review by kasperpeulen — APPROVED** [review]

**Comment by IanVS** [comment]
We will also want to update https://github.com/igor-dv/jest-specific-snapshot/pull/44 once it's merged and released, but I don't think it needs to block this PR.

File: code/addons/interactions/package.json
Comment (IanVS) [inline code/addons/interactions/package.json:85]: I don't think we can update this, since it's an actual dependency that is used by the interactions addon, and so far it only supports jest 27, AFAIK

Comment (kasperpeulen) [reply code/addons/interactions/package.json:85]: Aha, will downgrade, cc @yannbf

**Comment by SimenB** [comment]
🎉 

Is it on purpose this was skipped? https://github.com/storybookjs/storybook/blob/157f76adfe3d61ed2a318344d77e3effabbffd2c/code/addons/storyshots/storyshots-core/package.json#L54

File: code/addons/docs/package.json
Comment (IanVS) [inline code/addons/docs/package.json:88]: @kasperpeulen I see that you updated @jest/transform here and in storyshots.  Was that intentional?  This is a dependency that is used by addon-docs, and my storybook is now crashing with an error message coming from `jest-util` via `@jest/transform`.

```
Uncaught TypeError: Cannot read properties of undefined (reading 'isTTY')
    at ./node_modules/jest-util/build/isInteractive.js 
```

Seems to come from https://github.com/facebook/jest/blob/e7edb75f8357090714213f252c5aaaa9b3c29f5f/packages/jest-util/src/isInteractive.ts#L10.  We can't run that in the browser, since `process` isn't defined there.

Comment (SimenB) [reply code/addons/docs/package.json:88]: should be trivial to replace `process` with a fake object for browser usage, no?

Comment (IanVS) [reply code/addons/docs/package.json:88]: I'm sure we can deal with it, yes. I just like to keep user facing changes separate from internal test updates. 

@SimenB since you're here, is this something that changed from 28 to 29?  I took a quick look and couldn't immediately figure out why this would start to fail now in 29.  Maybe you know offhand?

Comment (IanVS) [reply code/addons/docs/package.json:88]: Hm, I upgraded to the latest beta (beta.8), and I no longer get this error.  I guess we can just keep our eyes out for it again.

Comment (SimenB) [reply code/addons/docs/package.json:88]: Nothing comes to mind, that code has been the same for years

**Pull request merged**
